### PR TITLE
[Flax] Dataset streaming example

### PR DESCRIPTION
**Pull request by patrickvonplaten**
# What does this PR do?

This PR adds an example of how to use dataset streaming.

@lhoestq - I would be super happy if you could take a look! I've left some comments that should help understand my thought process. I'd be very happy about some Feedback :-) 

The scripts runs as expected at the moment - after your feedback, I'll make it nicer (add better defaults) & add the eval coed as well :-)

### CODE REVIEW

File: examples/research_projects/jax-projects/dataset-streaming/run_mlm_flax_stream.py [hunk 1]
```diff
@@ -0,0 +1,652 @@
...
+        metadata={
+            "help": "Whether to pad all samples to `max_seq_length`. "
+            "If False, will pad the samples dynamically when batching to the maximum length in the batch."
+        },
+    )
+    line_by_line: bool = field(
+        default=False,
+        metadata={"help": "Whether distinct lines of text in the dataset are to be handled as distinct sequences."},
+    )
+    num_steps_per_stream_epoch: int = field(
```
Comment (patrickvonplaten) [inline examples/research_projects/jax-projects/dataset-streaming/run_mlm_flax_stream.py:181]: This is a new argument just for dataset streaming. My idea here is the following:
- I define one `"epoch"` as one time streaming `num_steps_per_stream_epoch` * `num_train_batch` samples
- => training 1 `epoch` here therefore means that `num_steps_per_stream_epoch` training steps have been finished. The default is super small here for testing, in reality it should be something like 500 - 1000 IMO (so that ~100_000 samples are streamed at once assuming a batch_size of ~128)

@lhoestq

File: examples/research_projects/jax-projects/dataset-streaming/run_mlm_flax_stream.py [hunk 2]
```diff
@@ -0,0 +1,652 @@
...
+    # (the dataset will be downloaded automatically from the datasets Hub).
+    #
+    # For CSV/JSON files, this script will use the column called 'text' or the first column if no column called
+    # 'text' is found. You can easily tweak this behavior (see below).
+    #
+    # In distributed training, the load_dataset function guarantees that only one local process can concurrently
+    # download the dataset.
+    if data_args.dataset_name is not None:
+        # Downloading and loading a dataset from the hub.
+        dataset = load_dataset(
```
Comment (patrickvonplaten) [inline examples/research_projects/jax-projects/dataset-streaming/run_mlm_flax_stream.py:348]: We just have a train set here -> I thought about adding a new variable `num_eval_batches` that defines how many batches from the original training set should be streamed for evaluation. @lhoestq

File: examples/research_projects/jax-projects/dataset-streaming/run_mlm_flax_stream.py [hunk 3]
```diff
@@ -0,0 +1,652 @@
...
+
+    model = FlaxAutoModelForMaskedLM.from_config(config, seed=training_args.seed, dtype=getattr(jnp, model_args.dtype))
+
+    # Store some constant
+    num_epochs = int(training_args.num_train_epochs)
+    train_batch_size = int(training_args.per_device_train_batch_size) * jax.device_count()
+    eval_batch_size = int(training_args.per_device_eval_batch_size) * jax.device_count()
+
+    # define number steps per stream epoch
+    num_steps_per_stream_epoch = data_args.num_steps_per_stream_epoch
```
Comment (patrickvonplaten) [inline examples/research_projects/jax-projects/dataset-streaming/run_mlm_flax_stream.py:478]: Here I define:

a) How many steps are in one stream epoch
b) How many total training steps are there?

File: examples/research_projects/jax-projects/dataset-streaming/run_mlm_flax_stream.py [hunk 4]
```diff
@@ -0,0 +1,652 @@
...
+
+    train_time = 0
+    #    epochs = tqdm(range(num_epochs), desc=f"Epoch ... (1/{num_epochs})", position=0)
+    num_train_samples_per_streaming = num_steps_per_stream_epoch * train_batch_size
+    stream_epoch = 0
+    next_stream_batch = None
+
+    # ======================== Training ================================
+    for step in range(num_train_steps):
+        if step % num_steps_per_stream_epoch == 0:
```
Comment (patrickvonplaten) [inline examples/research_projects/jax-projects/dataset-streaming/run_mlm_flax_stream.py:583]: This is an important part. Here I say that whenever all samples in one `stream_epoch` are used` we need to stream again

File: examples/research_projects/jax-projects/dataset-streaming/run_mlm_flax_stream.py [hunk 5]
```diff
@@ -0,0 +1,652 @@
...
+            model_inputs = data_collator(samples)
+
+            # Model forward
+            model_inputs = shard(model_inputs.data)
+            state, train_metric, dropout_rngs = p_train_step(state, model_inputs, dropout_rngs)
+            train_metrics.append(train_metric)
+
+        train_time += time.time() - train_start
+
+#        epochs.write(
```
Comment (patrickvonplaten) [inline examples/research_projects/jax-projects/dataset-streaming/run_mlm_flax_stream.py:609]: Still need to write this

Comment (patrickvonplaten) [reply examples/research_projects/jax-projects/dataset-streaming/run_mlm_flax_stream.py:609]: But should be easy

File: examples/research_projects/jax-projects/dataset-streaming/run_mlm_flax_stream.py [hunk 6]
```diff
@@ -0,0 +1,652 @@
...
+            train_start = time.time()
+            train_metrics = []
+
+            # Create sampling rng
+            rng, input_rng = jax.random.split(rng)
+
+            next_stream_batch = [next(iter(tokenized_datasets)) for _ in range(num_train_samples_per_streaming)]
+
+        # Generate an stream epoch by shuffling sampling indices from the train dataset
+        train_samples_idx = jax.random.permutation(input_rng, jnp.arange(num_train_samples_per_streaming))
```
Comment (patrickvonplaten) [inline examples/research_projects/jax-projects/dataset-streaming/run_mlm_flax_stream.py:594]: I shuffle within the streaming batch - think this should be good enough actually assuming that the streaming batch is big enough. Should we maybe shuffle the dataset right in the beginning after it's "stream" loaded as you explained here: https://huggingface.co/docs/datasets/dataset_streaming.html?highlight=streaming#shuffle-the-dataset @lhoestq ?

File: examples/research_projects/jax-projects/dataset-streaming/run_mlm_flax_stream.py [hunk 7]
```diff
@@ -0,0 +1,652 @@
...
+    state = jax_utils.replicate(state)
+
+    train_time = 0
+    #    epochs = tqdm(range(num_epochs), desc=f"Epoch ... (1/{num_epochs})", position=0)
+    num_train_samples_per_streaming = num_steps_per_stream_epoch * train_batch_size
+    stream_epoch = 0
+    next_stream_batch = None
+
+    # ======================== Training ================================
+    for step in range(num_train_steps):
```
Comment (patrickvonplaten) [inline examples/research_projects/jax-projects/dataset-streaming/run_mlm_flax_stream.py:582]:
```suggestion
    for step in range(num_train_steps):
          train_iter = iter(tokenized_datasets)
```

File: examples/research_projects/jax-projects/dataset-streaming/run_mlm_flax_stream.py [hunk 8]
```diff
@@ -0,0 +1,652 @@
...
+        if step % num_steps_per_stream_epoch == 0:
+            stream_epoch += 1
+            train_start = time.time()
+            train_metrics = []
+
+            # Create sampling rng
+            rng, input_rng = jax.random.split(rng)
+
+            next_stream_batch = [next(iter(tokenized_datasets)) for _ in range(num_train_samples_per_streaming)]
```
Comment (patrickvonplaten) [inline examples/research_projects/jax-projects/dataset-streaming/run_mlm_flax_stream.py:591]:
```suggestion
            next_stream_batch = [next(train_iter) for _ in range(num_train_samples_per_streaming)]
```

Comment (patrickvonplaten) [reply examples/research_projects/jax-projects/dataset-streaming/run_mlm_flax_stream.py:591]: @patrickvonplaten change to batch_size

File: examples/research_projects/jax-projects/dataset-streaming/run_mlm_flax_stream.py (hunk 6 above)
Comment (patrickvonplaten) [reply examples/research_projects/jax-projects/dataset-streaming/run_mlm_flax_stream.py:594]: Can be deleted here - and use official shuffle instead

File: examples/research_projects/jax-projects/dataset-streaming/run_mlm_flax_stream.py [hunk 9]
```diff
@@ -0,0 +1,622 @@
...
+    HfArgumentParser,
+    PreTrainedTokenizerBase,
+    TensorType,
+    TrainingArguments,
+    is_tensorboard_available,
+    set_seed,
+)
+
+
+if datasets.__version__ < "1.8.0":
```
Comment (patrickvonplaten) [inline examples/research_projects/jax-projects/dataset-streaming/run_mlm_flax_stream.py:61]:
```suggestion
if datasets.__version__ <= "1.8.0":
```

File: examples/research_projects/jax-projects/dataset-streaming/run_mlm_flax_stream.py [hunk 10]
```diff
@@ -0,0 +1,624 @@
...
+    set_seed(training_args.seed)
+
+    # Get the datasets: you can either provide your own CSV/JSON/TXT training and evaluation files (see below)
+    # or just provide the name of one of the public datasets available on the hub at https://huggingface.co/datasets/
+    # (the dataset will be downloaded automatically from the datasets Hub).
+    #
+    # For CSV/JSON files, this script will use the column called 'text' or the first column if no column called
+    # 'text' is found. You can easily tweak this behavior (see below).
+    #
+    # In distributed training, the load_dataset function guarantees that only one local process can concurrently
```
Comment (patil-suraj) [inline examples/research_projects/jax-projects/dataset-streaming/run_mlm_flax_stream.py:365]: We could delete this comment

File: examples/research_projects/jax-projects/dataset-streaming/run_mlm_flax_stream.py [hunk 11]
```diff
@@ -0,0 +1,624 @@
...
+
+    # Otherwise, we tokenize every text, then concatenate them together before splitting them in smaller parts.
+    # We use `return_special_tokens_mask=True` because DataCollatorForLanguageModeling (see below) is more
+    # efficient when it receives the `special_tokens_mask`.
+    def tokenize_function(examples):
+        return tokenizer(examples[data_args.text_column_name], return_special_tokens_mask=True)
+
+    tokenized_datasets = dataset.map(
+        tokenize_function,
+        batched=True,
```
Comment (patil-suraj) [inline examples/research_projects/jax-projects/dataset-streaming/run_mlm_flax_stream.py:407]: is it not possible to use `num_proc` here?

File: examples/research_projects/jax-projects/dataset-streaming/run_mlm_flax_stream.py [hunk 12]
```diff
@@ -0,0 +1,624 @@
...
+        flat_params = traverse_util.flatten_dict(params)
+        flat_mask = {path: (path[-1] != "bias" and path[-2:] != ("LayerNorm", "scale")) for path in flat_params}
+        return traverse_util.unflatten_dict(flat_mask)
+
+    # create adam optimizer
+    adamw = optax.adamw(
+        learning_rate=linear_decay_lr_schedule_fn,
+        b1=training_args.adam_beta1,
+        b2=training_args.adam_beta2,
+        eps=1e-8,
```
Comment (patil-suraj) [inline examples/research_projects/jax-projects/dataset-streaming/run_mlm_flax_stream.py:473]:
```suggestion
        eps=training_args.adam_epsilon,
```

**Comment by patrickvonplaten** [comment]
The training speed of this script compared to `run_mlm_flax` is sadly reduced by ~50%. @lhoestq - I think the "batchify" function that is called before every training step is probably the reason for this.

File: examples/research_projects/jax-projects/dataset-streaming/run_mlm_flax_stream.py [hunk 13]
```diff
@@ -0,0 +1,624 @@
...
+    )
+
+    steps = tqdm(range(num_train_steps), desc=f"Training...", position=0)
+    for step in range(num_train_steps):
+        # ======================== Training ================================
+        try:
+            training_iter, samples = advance_iter_and_group_samples(training_iter, train_batch_size, max_seq_length)
+        except StopIteration:
+            # Once the end of the dataset stream is reached, the training iterator
+            # is reinitialized and reshuffled and a new eval dataset is randomely chosen.
```
Comment (patil-suraj) [inline examples/research_projects/jax-projects/dataset-streaming/run_mlm_flax_stream.py:556]: is there any way to fix the eval dataset, changing the eval dataset would make it difficult to track the metrics no?

**Review by patil-suraj — APPROVED** [review]
Thanks for adding this! I left a few comments.

File: examples/research_projects/jax-projects/dataset-streaming/run_mlm_flax_stream.py (hunk 11 above)
Comment (patrickvonplaten) [reply examples/research_projects/jax-projects/dataset-streaming/run_mlm_flax_stream.py:407]: don't think the StreamingIterator has that cc @lhoestq

File: examples/research_projects/jax-projects/dataset-streaming/run_mlm_flax_stream.py (hunk 13 above)
Comment (patrickvonplaten) [reply examples/research_projects/jax-projects/dataset-streaming/run_mlm_flax_stream.py:556]: good point! IMO, this script should be used only for huge datasets where most of the times running for a single epoch is enough, so that this doesn't really matter. 

@lhoestq - what do you think?

File: examples/research_projects/jax-projects/dataset-streaming/run_mlm_flax_stream.py (hunk 11 above)
Comment (lhoestq) [reply examples/research_projects/jax-projects/dataset-streaming/run_mlm_flax_stream.py:407]: Not yet unfortunately ^^

File: examples/research_projects/jax-projects/dataset-streaming/run_mlm_flax_stream.py (hunk 13 above)
Comment (lhoestq) [reply examples/research_projects/jax-projects/dataset-streaming/run_mlm_flax_stream.py:556]: Once this PR is merged https://github.com/huggingface/datasets/pull/2582 you could do something like
```python
dataset = load_dataset(..., streaming=True).shuffle(...)

eval_size = 10000  # feel free to change the size
eval_dataset = dataset.take(eval_size)
train_dataset = dataset.skip(eval_size).shuffle(...)
```

File: examples/research_projects/jax-projects/dataset-streaming/run_mlm_flax_stream.py [hunk 14]
```diff
@@ -0,0 +1,624 @@
...
+    # Note that remainedr of % max_seq_length are thrown away.
+    def group_texts(examples):
+        result = {
+            k: [t[i : i + max_seq_length] for i in range(0, num_total_tokens, max_seq_length)]
+            for k, t in examples.items()
+        }
+        return result
+
+    grouped_samples = group_texts(samples)
+    return train_iterator, grouped_samples
```
Comment (lhoestq) [inline examples/research_projects/jax-projects/dataset-streaming/run_mlm_flax_stream.py:296]: I don't think you need to return the `train_iterator`

It is a stateful variable, so you return the same object as `train_iterator` passed in argument.
